qiskit/transpiler/target.py

```diff
@@ -350,8 +353,15 @@ def add_instruction(self, instruction, properties=None, name=None):
         if is_class:
             qargs_val = {None: None}
         else:
+            if None in properties:
+                self._global_operations[instruction.num_qubits].add(instruction_name)
             qargs_val = {}
             for qarg in properties:
+                if qarg is not None and len(qarg) != instruction.num_qubits:
+                    raise TranspilerError(
+                        f"The number of qubits for {instruction} does not match the number "
+                        f"of qubits in the properties dictionary: {qarg}"
+                    )
```

Comment:
This is probably out-of-scope for this PR, but some minor griping: the logic flow of this function is getting a little tricky for me now, since when I originally read this, I thought it would cause bugs with variadic instructions, as you can't do `type.num_qubits`.  In reality, it all works because `properties` is required to be `{None: None}`, so `qarg is not None` implies (transitively) that `instruction.num_qubits` should be valid, but because it's in a `for qarg in properties` _loop_, that's not so obvious to me any more.

Basically, I'm just griping that it's getting hard to keep all the different assumptions/implications about the objects in my head now.  There's nothing actionable unless there's an obvious way to refactor the logic a little now that we've been forced to support variadic arguments.

Reply:
Yeah, I feel like I need to write a big comment about all the different internal states and ways we track things. Part of it just is there are a lot of fiddly details here as there are basically 4 classes of instructions: normal instructions that have qubit restrictions either with or without properties, globally defined ideal instances of operations that have qubit and parameter restrictions, and globally defined variable width instructions. We've basically grafted the last two of those onto the target model which was designed for the first two (the 3rd class was originally just a shorthand to support ideal simulation without adding every permutation of qubit with no properties payload).

In this particular case I had tried to do this check outside the loop but I had logic flow issues with that as we still need to run this loop for the global fixed width instruction case even if this check itself isn't valid for those cases. I think we can look at trying to improve this in a separate refactor in a followup, especially as that I wouldn't want to backport necessarily.

Reply:
Yeah, it's absolutely a follow-up thing, if we do it at all.